python/ray/data/_internal/execution/interfaces/physical_operator.py

```diff
@@ -246,6 +242,11 @@ def add_input(self, refs: RefBundle, input_index: int) -> None:
                 input. For most operators, this is always `0` since there is only
                 one upstream input operator.
         """
+        self._metrics.on_input_received(refs)
+        self._add_input_inner(refs, input_index)
+
+    def _add_input_inner(self, refs: RefBundle, input_index: int) -> None:
+        """Subclasses should override this method to implement `add_input`."""
```

Comment:
should we also add a note in `add_input` docstring, that subclasses should really implement `_add_input_inner` instead? those unfamiliar with the codebase may not realize this from initially looking at the `add_input` method

also wondering, do we need to create the `_add_input_inner` method necessarily? could subclasses not call its `super()` `add_input()` and also have its own implementation inside its own `add_input`?

Reply:
> should we also add a note in add_input docstring, that subclasses should really implement _add_input_inner instead? those unfamiliar with the codebase may not realize this from initially looking at the add_input method

I didn't add this because I didn't want to mix "docstring for callers" and "docstring for subclasses". But after a second thought, I think this may not be a big deal. I will add it.

> also wondering, do we need to create the _add_input_inner method necessarily? could subclasses not call its super() add_input() and also have its own implementation inside its own add_input?

The issue with this method is that it's easy for subclasses to forgot call `super`, and unclear where to call `super` (at the beginning or the end). Having a separate method can make the interface clearer.

Reply:
Agree to add comment, I hope Python has @private, @public, @protected ...